### ISSUE TITLE
ActiveRecord::Base#create raises ArgumentError: string contains null byte

### ISSUE THREAD

**Issue by sonalkr132** [rails/rails#26891]
### Steps to reproduce

Run following:

``` Ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  # Activate the gem you are reporting the issue against.
  gem "activerecord", "4.2.7.1"
  gem "pg"
end

require "active_record"
require "minitest/autorun"
require "logger"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(
  adapter: 'postgresql',
  database: 'myapp_test',
  username: <username>,
  password: <password>,
  host: 'localhost',
  pool: 5,
  timeout: 5000
)

ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.string :title
  end
end

class Post < ActiveRecord::Base
end

title = "1<svg\t\n\v\f\r�\u0000onload=7zvO(9278)"
Post.create(title: title)
```
### Expected behavior

Post should have been created.
### Actual behavior

> contains null byte (ArgumentError)
>     from /home/user/.rvm/gems/ruby-2.3.1/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/postgresql_adapter.rb:602:in `block in exec_cache'
>     from /home/user/.rvm/gems/ruby-2.3.1/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract_adapter.rb:484:in`block in log'
>     from /home/user/.rvm/gems/ruby-2.3.1/gems/activesupport-4.2.7.1/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
>     from /home/user/.rvm/gems/ruby-2.3.1/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/abstract_adapter.rb:478:in`log'
>     from /home/user/.rvm/gems/ruby-2.3.1/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/postgresql_adapter.rb:601:in `exec_cache'
>     from /home/user/.rvm/gems/ruby-2.3.1/gems/activerecord-4.2.7.1/lib/active_record/connection_adapters/postgresql_adapter.rb:585:in`execute_and_clear'
> ...
### System configuration

**Rails version**:
4.2.7.1
**Ruby version**:
ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-linux]

Related: #8957, #23035

**Comment by maclover7** [comment]
Fails on master as well.

**Comment by why-el** [comment]
The PG gem uses a ruby macro to try and ensure your strings are properly null terminated, and if not will raise that error you see. This will work the same if you use a `where` for example, and more generally when a binary value is tried against non-binary column types. I am not sure if Rails should mingle in this, since it fails just the same if you mismatches for many other types.

**Comment by rafaelfranca** [comment]
Yes. This is not an issue.

**Comment by leoplct** [comment]
For me It was caused by a string containing "\u0000"

**Comment by doviende** [comment]
For future reference, we ran into this on Rails 4.2.11 / pg 0.20 / ruby 2.2, using a bytea column to store binary data. If we build a new object (with a null byte inside the binary field) and save, then dup will raise `ArgumentError` as mentioned above. During the `dup`, as it is trying to clear the timestamp fields, we notice that the binary attribute's `value` and `value_before_type_cast` are both already unescaped and both strings, so the given fix of checking for `Type::Binary::Data` doesn't apply, and it tries to unescape a second time. Doing a `.reload` prevents the problem. Setting the `created_at` and `updated_at` to nil before duping also prevents the problem, at least in the specific case of dup.

adding this seems to fix it, by trying to detect whether the string is unescaped before sending it to unescape_bytea:

         return value.to_s if value.is_a?(Type::Binary::Data)
         return value unless value.slice(0,2) == "\\x"
         PGconn.unescape_bytea(super)
(not sure if that's the most general solution, and it doesn't address the cause of why the data is already unescaped, but I figured I'd convey the information for whoever may need it)

problem is gone in Rails 5+
